### PR TITLE
nixos/tests/boot: fix intermittent biosUsb failure

### LINKED ISSUE #15690: Random test failures in nixos.tests.boot.biosUsb
http://hydra.nixos.org/build/36146783:

```
machine# ISOLINUX 6.04   Copyright (C) 1994-2015 H. Peter Anvin et al
machine# CHS: Error 0c00 reading sector 4220 (4/2/63)
machine# EDD: Error 0c00 reading sector 4220
error: timed out waiting for the VM to connect
```

Note: error 0c00 = unsupported track or invalid media (http://www.ctyme.com/intr/rb-0606.htm#Table234)

http://hydra.nixos.org/build/36146731 reaches the syslinux menu, then hangs loading the initrd:

```
Loading /boot/bzImage... ok
error: timed out waiting for the VM to connect
```

http://hydra.nixos.org/build/35988622:

```
machine# ISOLINUX 6.04   Copyright (C) 1994-2015 H. Peter Anvin et al
error: timed out waiting for the VM to connect
```

http://hydra.nixos.org/build/36146446:

```
Loading /boot/bzImage... CHS: Error 0c00 reading sector 11339 (11/3/63)
machine# EDD: Error 0c00 reading sector 11339
error: timed out waiting for the VM to connect
```

Haven't been able to reproduce these locally, so it might be triggered by high load on the host.

This may be a QEMU USB emulation issue. It doesn't appear to affect the USB UEFI boot test though.

### LINKED ISSUE #104642: vm-test-run-boot-bios-usb blocker
**Describe the bug**

Build failure on hydra release-20.09
https://hydra.nixos.org/build/131059528

([blocker](https://hydra.nixos.org/build/131060533#tabs-summary))

**To Reproduce**

https://hydra.nixos.org/build/131059528

**Expected behavior**

**Screenshots**

**Additional context**

**Notify maintainers**

@edolstra @fpletz 

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
 - nixosTests.boot.biosUsb
# a list of nixos modules affected by the problem
module:
```

### LINKED ISSUE #170803: nixos.tests.boot.biosUsb.x86_64-linux fails on hydra: "cpage out of range (5)"
https://hydra.nixos.org/build/174964149

```
machine # Booting from Hard Disk...
machine # 
machine # ISOLINUX 6.04  EHDD Copyright (C) 1994-2015 H. Peter Anvin et a
machine # ISOLINUX 6.04   Copyright (C) 1994-2015 H. Peter Anvin et al
machine # l
machine # e%@)0(Blqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqkx                                             NixOS                                              xtqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqux NixOS 22.05pre-git Installer                                                                   xx NixOS 22.05pre-git Installer (nomodeset)                                                       xx NixOS 22.05pre-git Installer (copytoram)                                                       xx NixOS 22.05pre-git Installer (debug)                                                           xx NixOS 22.05pre-git Installer (serial console=ttyS0,115200n8)                                   xx Memtest86+                                                                                     xmqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqjPress [Tab] to edit optionsAutomatic boot in 10 seconds... Automatic boot in 9 seconds... Automatic boot in 8 seconds...Automatic boot in 7 seconds...Automatic boot in 6 seconds...Automatic boot in 5 seconds...Automatic boot in 4 seconds...Automatic boot in 3 seconds...Automatic boot in 2 seconds... Automatic boot in 1 second... lqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqkx                                             NixOS                                              xtqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqux NixOS 22.05pre-git Installer                                                                   xx NixOS 22.05pre-git Installer (nomodeset)                                                       xx NixOS 22.05pre-git Installer (copytoram)                                                       xx NixOS 22.05pre-git Installer (debug)                                                           xx NixOS 22.05pre-git Installer (serial console=ttyS0,115200n8)                                   xx Memtest86+                                                                                     xmqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqjPress [Tab] to edit optionse%@)0(BLoading /boot/bzImage... cpage out of range (5)
machine # processing error - resetting ehci HC
machine # CHS: Error 0c00 reading sector 3058 (1/16/35)
machine # EDD: Error 0c00 reading sector 3058
machine # CHS: Error 0c00 reading sector 15543 (7/22/46)
machine # EDD: Error 0c00 reading sector 15543
machine # ok
machine # Loading /boot/initrd...CHS: Error 0c00 reading sector 70844 (35/4/33)
machine # EDD: Error 0c00 reading sector 70844
machine # CHS: Error 0c00 reading sector 96311 (47/24/48)
machine # EDD: Error 0c00 reading sector 96311
machine # ok
machine: connected to guest root shell
machine: (connecting took 1017.66 seconds)
(finished: waiting for the VM to finish booting, in 1017.66 seconds)
cleanup
(finished: cleanup, in 0.00 seconds)
Traceback (most recent call last):
  File "/nix/store/7m0m07v3yhj81l2p1sbj8krlzximmd21-nixos-test-driver-1.1/bin/.nixos-test-driver-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/7m0m07v3yhj81l2p1sbj8krlzximmd21-nixos-test-driver-1.1/lib/python3.9/site-packages/test_driver/__init__.py", line 114, in main
    driver.run_tests()
  File "/nix/store/7m0m07v3yhj81l2p1sbj8krlzximmd21-nixos-test-driver-1.1/lib/python3.9/site-packages/test_driver/driver.py", line 146, in run_tests
    self.test_script()
  File "/nix/store/7m0m07v3yhj81l2p1sbj8krlzximmd21-nixos-test-driver-1.1/lib/python3.9/site-packages/test_driver/driver.py", line 142, in test_script
    exec(self.tests, symbols, None)
  File "<string>", line 9, in <module>
  File "/nix/store/7m0m07v3yhj81l2p1sbj8krlzximmd21-nixos-test-driver-1.1/lib/python3.9/site-packages/test_driver/machine.py", line 459, in wait_for_unit
    retry(check_active)
  File "/nix/store/7m0m07v3yhj81l2p1sbj8krlzximmd21-nixos-test-driver-1.1/lib/python3.9/site-packages/test_driver/machine.py", line 126, in retry
    if fn(False):
  File "/nix/store/7m0m07v3yhj81l2p1sbj8krlzximmd21-nixos-test-driver-1.1/lib/python3.9/site-packages/test_driver/machine.py", line 436, in check_active
    info = self.get_unit_info(unit, user)
  File "/nix/store/7m0m07v3yhj81l2p1sbj8krlzximmd21-nixos-test-driver-1.1/lib/python3.9/site-packages/test_driver/machine.py", line 462, in get_unit_info
    status, lines = self.systemctl('--no-pager show "{}"'.format(unit), user)
  File "/nix/store/7m0m07v3yhj81l2p1sbj8krlzximmd21-nixos-test-driver-1.1/lib/python3.9/site-packages/test_driver/machine.py", line 493, in systemctl
    return self.execute("systemctl {}".format(q))
  File "/nix/store/7m0m07v3yhj81l2p1sbj8krlzximmd21-nixos-test-driver-1.1/lib/python3.9/site-packages/test_driver/machine.py", line 541, in execute
    self.shell.send(out_command.encode())
BrokenPipeError: [Errno 32] Broken pipe
machine # cProbing EDD (edd=off to disable)... oc
```

Specifically:

```
Loading /boot/bzImage... cpage out of range (5)
```

For comparison, a successful run loads bzImage successfully:

```
machine # ISOLINUX 6.04  EHDD Copyright (C) 1994-2015 H. Peter Anvin et a
machine # ISOLINUX 6.04   Copyright (C) 1994-2015 H. Peter Anvin et al
machine # l
machine # e%@)0(Blqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqkx                                             NixOS                                              xtqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqux NixOS 22.05pre-git Installer                                                                   xx NixOS 22.05pre-git Installer (nomodeset)                                                       xx NixOS 22.05pre-git Installer (copytoram)                                                       xx NixOS 22.05pre-git Installer (debug)                                                           xx NixOS 22.05pre-git Installer (serial console=ttyS0,115200n8)                                   xx Memtest86+                                                                                     xmqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqjPress [Tab] to edit optionsAutomatic boot in 10 seconds... Automatic boot in 9 seconds... Automatic boot in 8 seconds...Automatic boot in 7 seconds...Automatic boot in 6 seconds...Automatic boot in 5 seconds...Automatic boot in 4 seconds...Automatic boot in 3 seconds...Automatic boot in 2 seconds... Automatic boot in 1 second... lqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqkx                                             NixOS                                              xtqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqux NixOS 22.05pre-git Installer                                                                   xx NixOS 22.05pre-git Installer (nomodeset)                                                       xx NixOS 22.05pre-git Installer (copytoram)                                                       xx NixOS 22.05pre-git Installer (debug)                                                           xx NixOS 22.05pre-git Installer (serial console=ttyS0,115200n8)                                   xx Memtest86+                                                                                     xmqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqjPress [Tab] to edit optionse%@)0(BLoading /boot/bzImage... ok
machine # Loading /boot/initrd...ok
```

### PR DESCRIPTION
**Pull request by ncfavier**
Fixes https://github.com/NixOS/nixpkgs/issues/15690, fixes https://github.com/NixOS/nixpkgs/issues/104642, fixes https://github.com/NixOS/nixpkgs/issues/170803 🤞

Based on our observations in https://github.com/NixOS/nixpkgs/issues/170803, the issue doesn't seem to occur when using the ACPI PM timer instead of the TSC timer, so we use a patched seabios for the biosUsb test with `CONFIG_TSC_TIMER` disabled.

Note that the bug is reported upstream https://mail.coreboot.org/hyperkitty/list/seabios@seabios.org/thread/OUTHT5ISSQJGXPNTUPY3O5E5EPZJCHM3/

### CODE REVIEW

File: nixos/tests/boot.nix [hunk 1]
```diff
@@ -114,8 +114,28 @@ in {
       cdrom = "${iso}/iso/${iso.isoName}";
     };
 
-    biosUsb = makeBootTest "bios-usb" {
+    biosUsb = let
+      seabiosNoTSC = pkgs.seabios.overrideAttrs (_: {
+        configurePhase = ''
+          cat > .config << EOF
```
Comment (raboof) [inline nixos/tests/boot.nix:120]: I think it would be neat to add 'tsc' and 'csm' parameters to the `seabios` derivation, but this works as well :)

**Review by raboof — APPROVED** [review]
I'm OK with this workaround, since testing the seabios/qemu integration is not the goal of this test.

We mainly looked at #170803 when we found out it's likely the TSC timer that triggers the problem, but indeed it's possible that the other problems have the same origin - I'm OK with closing those now and re-opening when it turns out problems remain.

I haven't tested that disabling TSC with this seabios configuration option indeed works around the issue.

File: nixos/tests/boot.nix (hunk 1 above)
Comment (ncfavier) [reply nixos/tests/boot.nix:120]: Good idea

**Comment by ncfavier** [comment]
@ofborg test boot.biosUsb

**Comment by raboof** [comment]
Perhaps no longer needed if we accept #172070

File: pkgs/applications/virtualization/seabios/default.nix [hunk 2]
```diff
@@ -1,4 +1,11 @@
-{ lib, stdenv, fetchgit, acpica-tools, python3 }:
+{ lib, stdenv, fetchgit, acpica-tools, python3
```
Comment (raboof) [inline pkgs/applications/virtualization/seabios/default.nix:1]: (the changes to this file might still be kinda nice to have, but no big deal anymore I guess)

**Pull request closed without merging**